osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/firmware/firmwarefile/FirmwareFile.java

```diff
@@ -0,0 +1,129 @@
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.firmware.firmwarefile;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import lombok.extern.slf4j.Slf4j;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.ProtocolAdapterException;
+
+/**
+ * This Firmware file is defined according to the Smart Meter Requirements 5.1, Supplement 4, P2
+ * Companion Standard, Section 7. Firmware Upgrade
+ *
+ * <p>Fixed values are
+ *
+ * <ul>
+ *   <li>Firmware image magic number: 53h 4Dh 52h 35h
+ *   <li>Header Version: 00h
+ */
+@Slf4j
+public class FirmwareFile {
+
+  private byte[] imageData;
+
+  public static final String FIRMWARE_IMAGE_MAGIC_NUMBER = "534d5235";
+  public static final int HEADER_VERSION = 0;
+
+  public FirmwareFile(final byte[] imageData) {
+    this.imageData = imageData;
+    final Integer firmwareImageLength = this.getHeader().getFirmwareImageLengthInt();
+    final Integer securityLength = this.getHeader().getSecurityLengthInt();
+    final Integer headerLength = this.getHeader().getHeaderLengthInt();
+    if (imageData.length != (firmwareImageLength + securityLength + headerLength)) {
+      log.warn(
+          "Byte array length doesn't match lengths defined in header: "
+              + "\nByte array length : {}"
+              + "\nLengths defined in header: "
+              + "\nHeader : {}"
+              + "\nFirmwareImage : {}"
+              + "\nSecurity : {}"
+              + "\nTotal of {}  bytes.",
+          imageData.length,
+          headerLength,
+          firmwareImageLength,
+          securityLength,
+          (firmwareImageLength + securityLength + headerLength));
+    }
+  }
+
+  public int length() {
+    return this.imageData.length;
+  }
+
+  public byte[] getByteArray() {
+    return this.imageData;
+  }
+
+  public byte[] getHeaderByteArray() {
+    return this.readBytes(this.imageData, 0, this.getHeader().getHeaderLengthInt() - 1);
+  }
+
+  public byte[] getFirmwareImageByteArray() {
+    return this.readBytes(
+        this.imageData,
+        this.getHeader().getHeaderLengthInt(),
+        this.getHeader().getHeaderLengthInt() + this.getHeader().getFirmwareImageLengthInt() - 1);
+  }
+
+  public void setSecurityByteArray(final byte[] security) throws ProtocolAdapterException {
+    if (security.length != this.getHeader().getSecurityLengthInt()) {
+      throw new ProtocolAdapterException(
+          "Length of generated security byte array differs from length defined in header of firmwarefile");
+    }
+    this.imageData =
+        ByteBuffer.allocate(this.length())
+            .put(this.getHeaderByteArray())
+            .put(this.getFirmwareImageByteArray())
+            .put(security)
+            .array();
+  }
+
+  public void setMbusDeviceSerialNumber(final int mbusDeviceSerialNumber) {
```

Comment:
The SMR 5.1 spec says both the identification number (set by this method) and the version number from the M-Bus address could be wildcarded. Should there also be a setter for the version or is that unused in this implementation?

Reply:
The version is a preset by the manufacturer and does not have to be set here